### PR TITLE
Feature/user defaults property wrapper

### PR DESCRIPTION
**Pull request by jabou**
# Description

Please include a summary of the feature you have added.

# Checklist:

- [x] Have checked there is no same component already in catalog.
- [x] Have created a sufficient example or wrote tests for it.
- [x] Code is structured, well written using standard Swift coding style.
- [x] All public methods and properties have meaningful and concise documentation.
- [x] Used `public` modifier for all method and properties that could be changed from user of your feature, and `private` for internal properties.
- [x] Removed any reference to the project that piece of code was created in.
- [x] Reduced the dependencies to minimum.

### CODE REVIEW

File: Sources/Foundation/UserDefaults/UserDefaults+PropertyWrapper.swift [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+    init<T>(codable key: String, userDefaults: UserDefaults = .standard, encoder: JSONEncoder = JSONEncoder(), decoder: JSONDecoder = JSONDecoder()) where Value == T? {
+        self.init(
+            key,
+            defaultValue: nil,
+            userDefaults: userDefaults,
+            get: {
+                guard let storedData = $0 else { return nil }
+                return try? decoder.decode(Value.self, from: storedData)
+        },
```
Comment (goranbrl) [inline Sources/Foundation/UserDefaults/UserDefaults+PropertyWrapper.swift:108]: Indentation 🙂

File: Sources/Foundation/UserDefaults/UserDefaults+PropertyWrapper.swift [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    init(codable key: String, defaultValue: Value, userDefaults: UserDefaults = .standard, encoder: JSONEncoder = JSONEncoder(), decoder: JSONDecoder = JSONDecoder()) {
+        self.init(
+            key,
+            defaultValue: defaultValue,
+            get: {
+                guard let storedData = $0 else {
+                    return defaultValue
+                }
+                return (try? decoder.decode(Value.self, from: storedData)) ?? defaultValue
+        },
```
Comment (goranbrl) [inline Sources/Foundation/UserDefaults/UserDefaults+PropertyWrapper.swift:95]: Here as well.

File: Tests/Foundation/UserDefaults/UserDefaults+PropertyWrapperTests.swift [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+//  Created by Jasmin Abou Aldan on 01/09/2020.
+//  Copyright © 2020 Infinum. All rights reserved.
+//
+
+import XCTest
+import Quick
+import Nimble
+@testable import Catalog
+
+class UserDefaults_PropertyWrapperTests: QuickSpec {
```
Comment (goranbrl) [inline Tests/Foundation/UserDefaults/UserDefaults+PropertyWrapperTests.swift:14]: 👌

**Review by goranbrl — APPROVED** [review]

File: Sources/Foundation/UserDefaults/UserDefaults+PropertyWrapper.swift [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    }
+}
+
+extension Storage where Value: RawRepresentable, Value.RawValue == StoredValue {
+
+    init(_ key: String, defaultValue: Value, userDefaults: UserDefaults = .standard) {
+        self.init(
+            key,
+            defaultValue: defaultValue,
+            get: { return Value.init(rawValue: $0) ?? defaultValue },
```
Comment (fgulan) [inline Sources/Foundation/UserDefaults/UserDefaults+PropertyWrapper.swift:56]: Init not needed here

File: Tests/Foundation/UserDefaults/UserDefaults+PropertyWrapperTests.swift [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+
+    override func spec() {
```
Comment (fgulan) [inline Tests/Foundation/UserDefaults/UserDefaults+PropertyWrapperTests.swift:16]: And yet two cases with most logic are not covered :( please test out RawRepresentable init and Codable inits.

Comment (jabou) [reply Tests/Foundation/UserDefaults/UserDefaults+PropertyWrapperTests.swift:16]: I have added test cases and minor fixes for Codable and RawRepresentable (non optional). I'll get back to this PR later to add support for optional RawRepresentable

Comment (jabou) [reply Tests/Foundation/UserDefaults/UserDefaults+PropertyWrapperTests.swift:16]: As agreed, optional RawRepresentable will not be supported for now :)

**Review by fgulan — APPROVED** [review]

**Pull request merged**